source/grid/tria_accessor.cc

```diff
@@ -1286,42 +1286,35 @@ namespace
 
 
   // a 2d face in 3d space
+  template <int dim>
   double
-  measure(const dealii::TriaAccessor<2, 3, 3> &accessor)
+  measure(const dealii::TriaAccessor<2, dim, 3> &accessor)
   {
-    // If the face is planar, the diagonal from vertex 0 to vertex 3,
-    // v_03, should be in the plane P_012 of vertices 0, 1 and 2.  Get
-    // the normal vector of P_012 and test if v_03 is orthogonal to
-    // that. If so, the face is planar and computing its area is simple.
-    const Tensor<1, 3> v01 = accessor.vertex(1) - accessor.vertex(0);
-    const Tensor<1, 3> v02 = accessor.vertex(2) - accessor.vertex(0);
-
-    const Tensor<1, 3> normal = cross_product_3d(v01, v02);
-
-    const Tensor<1, 3> v03 = accessor.vertex(3) - accessor.vertex(0);
-
-    // check whether v03 does not lie in the plane of v01 and v02
-    // (i.e., whether the face is not planar). we do so by checking
-    // whether the triple product (v01 x v02) * v03 forms a positive
-    // volume relative to |v01|*|v02|*|v03|. the test checks the
-    // squares of these to avoid taking norms/square roots:
-    if (std::abs((v03 * normal) * (v03 * normal) /
-                 ((v03 * v03) * (v01 * v01) * (v02 * v02))) >= 1e-24)
-      {
-        Assert(
-          false,
-          ExcMessage(
-            "Computing the measure of a nonplanar face is not implemented!"));
-        return std::numeric_limits<double>::quiet_NaN();
-      }
+    // In general the area can be computed as
+    // 0.25*(v_0+v_1-v_2-v_3)*(v_0-v_1+v_2-v_3)
```

Comment:
That's interesting. Can you explain why this is so? I have to say that geometrically this formula seems unlikely to me because you are, in essence taking the cross product of the tangent vectors at the center.

Think about the planar face in 3d with coordinates
```
  0,0,0      1,0,0     0,1,0     1,1,0
```
and then the one where you lift opposing vertices above the xy plane and move the others below:
```
  0,0,1      1,0,-1     0,1,-1     1,1,1
```
Your formula predicts the same are, but that's clearly not the case.

Reply:
You are absolutely right again. Thanks, I will look again tomorrow, and If I don't found the mistake I will close the PR.

Reply:
The area you compute is that of the projection onto the plane perpendicular to the normal vector of the face at the center of the face. I would not be surprised if there is a formula that expresses the true area in terms of a warp or twist relative to this projection.

But think about how you would compute things mathematically: You want to compute
```
  \int_F  1   ds_1 ds_2
```
which you can pull back to the reference face using
```
  \int_{\hat F}   det(ds/d\hat s)  dxi deta
  = \int_{\hat F}   det(J)  dxi deta
```
The Jacobian (of the mapping) can be computed from the coordinates of the face. Then the integrand will be of the form `polynomial of degree d` and should be computable. The only thing you'll have to pay attention to is that `J` is a `3x2` matrix, and I think you have to replace `det(J)` by `sqrt(det(J^T J))`. I'll leave the details for you to figure out :-)

Reply:
The surface in codim one should be defined as the integral of the norm of the cross product of the two tangential vectors. I have assumed a bi-linear patch, am I correct to assume that the proper result should be the same as the one computed integrating 1 on the patch using a MappingQ1 and a properly accurate QGauss?

Reply:
I think that's true. The norm of the cross product of the two properly scaled tangential vectors is simply the determinant of the Jacobian.

The current formula just makes use that under certain situations, the integrand is constant and so can be computed analytically.